From 10c6725d8a64db406cb54f7a86eadf19ae8789da Mon Sep 17 00:00:00 2001
From: Francois Lallart <fraff@free.fr>
Date: Mon, 19 May 2014 12:05:55 +0200
Subject: [PATCH] The "Date:" mail header is now localtime.

---
 keepalived/core/smtp.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/keepalived/core/smtp.c b/keepalived/core/smtp.c
index 6b1cf7e..b6b3c7e 100644
--- a/keepalived/core/smtp.c
+++ b/keepalived/core/smtp.c
@@ -485,11 +485,13 @@ body_cmd(thread_t * thread)
 	char *buffer;
 	char rfc822[80];
 	time_t tm;
+	struct tm *t;
 
 	buffer = (char *) MALLOC(SMTP_BUFFER_MAX);
 
 	time(&tm);
-	strftime(rfc822, sizeof(rfc822), "%a, %d %b %Y %H:%M:%S %z", gmtime(&tm));
+	t = localtime(&tm);
+	strftime(rfc822, sizeof(rfc822), "%a, %d %b %Y %H:%M:%S %z", t);
 
 	snprintf(buffer, SMTP_BUFFER_MAX, SMTP_HEADERS_CMD,
 		 rfc822, global_data->email_from, smtp->subject, smtp->email_to);
-- 
1.8.3.2

