From 9d701d95a8e93ebfa773149f22d6965883597fd1 Mon Sep 17 00:00:00 2001
From: Gu Feng <flygoast@126.com>
Date: Thu, 20 Aug 2015 18:07:03 +0800
Subject: [PATCH 16/17] use getaddrinfo() instread of gethostbyname()

---
 lib/utils.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/lib/utils.c b/lib/utils.c
index b2a4a9d..b6b4cb9 100644
--- a/lib/utils.c
+++ b/lib/utils.c
@@ -169,24 +169,23 @@ inet_stor(char *addr)
 int
 domain_stosockaddr(char *domain, char *port, struct sockaddr_storage *addr)
 {
-	struct hostent *h = gethostbyname(domain);
+	struct addrinfo *res = NULL;
 
-	if (!h  || !(h->h_addr_list[0]))
+	if (getaddrinfo(domain, NULL, NULL, &res) != 0 || !res)
 		return -1;
 
-	addr->ss_family = h->h_addrtype;
+	addr->ss_family = res->ai_family;
 
 	if (addr->ss_family == AF_INET6) {
 		struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *) addr;
+		*addr6 = *(struct sockaddr_in6 *) res->ai_addr;
 		if (port)
 			addr6->sin6_port = htons(atoi(port));
-
-		addr6->sin6_addr = *(struct in6_addr *) h->h_addr_list[0];
 	} else {
 		struct sockaddr_in *addr4 = (struct sockaddr_in *) addr;
+		*addr4 = *(struct sockaddr_in *) res->ai_addr;
 		if (port)
 			addr4->sin_port = htons(atoi(port));
-		addr4->sin_addr = *(struct in_addr *) h->h_addr_list[0];
 	}
 
 	return 0;
-- 
2.4.6

