From 249423421bb3b45120bc656906cdc3b9d95765ea Mon Sep 17 00:00:00 2001
From: Gu Feng <flygoast@126.com>
Date: Thu, 20 Aug 2015 18:33:43 +0800
Subject: [PATCH 17/17] fixed error of forgotten 'freeaddrinfo()'

---
 lib/utils.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/utils.c b/lib/utils.c
index b6b4cb9..8535aea 100644
--- a/lib/utils.c
+++ b/lib/utils.c
@@ -188,6 +188,8 @@ domain_stosockaddr(char *domain, char *port, struct sockaddr_storage *addr)
 			addr4->sin_port = htons(atoi(port));
 	}
 
+	freeaddrinfo(res);
+
 	return 0;
 }
 
-- 
2.4.6

