Index: gcc-4.6.2/libstdc++-v3/include/precompiled/stdtr1c++.h
===================================================================
--- gcc-4.6.2.orig/libstdc++-v3/include/precompiled/stdtr1c++.h
+++ gcc-4.6.2/libstdc++-v3/include/precompiled/stdtr1c++.h
@@ -30,7 +30,6 @@
 
 #include <tr1/array>
 #include <tr1/cctype>
-#include <tr1/cfenv>
 #include <tr1/cfloat>
 #include <tr1/cinttypes>
 #include <tr1/climits>
Index: gcc-4.6.2/libstdc++-v3/include/precompiled/stdc++.h
===================================================================
--- gcc-4.6.2.orig/libstdc++-v3/include/precompiled/stdc++.h
+++ gcc-4.6.2/libstdc++-v3/include/precompiled/stdc++.h
@@ -50,8 +50,6 @@
 #include <ctime>
 
 #ifdef __GXX_EXPERIMENTAL_CXX0X__
-#include <ccomplex>
-#include <cfenv>
 #include <cinttypes>
 #include <cstdbool>
 #include <cstdint>
