Index: iptables-mod-ndpi-1.2/src/include/ndpi_protocols_osdpi.h
===================================================================
--- iptables-mod-ndpi-1.2.orig/src/include/ndpi_protocols_osdpi.h
+++ iptables-mod-ndpi-1.2/src/include/ndpi_protocols_osdpi.h
@@ -36,6 +36,42 @@ extern "C" {
 
 //  #define NDPI_ENABLE_DEBUG_MESSAGES
 
+#define NDPI_PROTOCOL_LONG_STRING "Unknown","FTP","POP","SMTP","IMAP","DNS","IPP","HTTP","MDNS","NTP","NETBIOS",\
+"NFS","SSDP","BGP","SNMP","XDMCP","SMB","SYSLOG","DHCP","PostgreSQL","MySQL",\
+"TDS","DirectDownloadLink","I23V5","AppleJuice","DirectConnect","Socrates","WinMX","MANOLITO","PANDO","Filetopia",\
+"iMESH","Kontiki","OpenFT","Kazaa","Gnutella","eDonkey","Bittorrent","OFF","AVI","Flash",\
+"OGG","MPEG","QuickTime","RealMedia","Windowsmedia","MMS","XBOX","QQ","MOVE","RTSP",\
+"Feidian","Icecast","PPLive","PPStream","Zattoo","SHOUTCast","SopCast","TVAnts","TVUplayer","VeohTV",\
+"QQLive","Thunder","Soulseek","GaduGadu","IRC","Popo","Jabber","MSN","Oscar","Yahoo",\
+"Battlefield","Quake","Second_Life","Steam","Halflife2","World_of_Warcraft","Telnet","STUN","IPSEC","GRE",\
+"ICMP","IGMP","EGP","SCTP","OSPF","IP_in_IP","RTP","RDP","VNC","PCAnywhere",\
+"SSL","SSH","USENET","MGCP","IAX","TFTP","AFP","StealthNet","Aimini","SIP",\
+"Truphone","ICMPv6","DHCPv6","Armagetron","CrossFire","Dofus","Fiesta","Florensia","Guildwars","HTTP_Application_Activesync",\
+"Kerberos","LDAP","MapleStory","msSQL","PPTP","WARCRAFT3","World_of_Kung_Fu","MEEBO","FaceBook","Twitter",\
+"DropBox","Gmail","Google_Maps","YouTube","Skype","Google","DCE_RPC","NetFlow_IPFIX","sFlow","HTTP_Connect_SSL_over_HTTP",\
+"HTTP_Proxy","Citrix","Netflix","LastFM","Grooveshark","SkyFile_Prepaid","SkyFile_Rudics","SkyFile_Postpaid","CitrixOnline","Apple_iMessage_FaceTime",\
+"Webex","WhatsApp","Apple_iCloud","Viber","Apple_iTunes","Radius","Windows_Update","TeamViewer","Tuenti","Lotus_Notes",\
+"SAP","GTP","uPnP","LLMNR","RemoteScan","Spotify","H323","OpenVPN","NOE","CiscoVPN","TeamSpeak","Tor","RTCP","RSYNC","Oracle","Corba","UbuntuONE"
+
+
+#define NDPI_PROTOCOL_SHORT_STRING "ukn","ftp","pop","smtp","imap","dns","ipp","http","mdns","ntp","netbios",\
+"nfs","ssdp","bgp","snmp","xdmcp","smb","syslog","dhcp","postgres","mysql",\
+"tds","ddl","i23v5","applejuice","directconnect","socrates","winmx","manolito","pando","filetopia",\
+"iMESH","kontiki","openft","fasttrack","gnutella","edonkey","bittorrent","off","avi","flash",\
+"ogg","mpeg","quicktime","realmedia","windowsmedia","mms","xbox","qq","move","rtsp",\
+"feidian","icecast","pplive","ppstream","zattoo","shoutcast","sopcast","tvants","tvuplayer","veohtv",\
+"qqlive","thunder","soulseek","gadugadu","irc","popo","jabber","msn","oscar","yahoo",\
+"battlefield","quake","secondlife","steam","hl2","worldofwarcraft","telnet","stun","ipsec","gre",\
+"icmp","igmp","egp","sctp","ospf","ipip","rtp","rdp","vnc","pcanywhere",\
+"ssl","ssh","usenet","mgcp","iax","tftp","afp","stealthnet","aimini","sip",\
+"truphone","icmpv6","dhcpv6","armagetron","crossfire","dofus","fiesta","florensia","guildwars","httpactivesync",\
+"kerberos","ldap","maplestory","mssql","pptp","warcraft3","wokf","meebo","facebook","twitter",\
+"dropbox","gmail","gmaps","youtube","skype","google","dcerpc","netflow","sflow","httpconnect",\
+"httpproxy","citrix","netflix","lastfm","grooveshark","skyfileprepaid","skyfilerudics","skyfilepostpaid","citrixonline","apple",\
+"webex","wgatsapp","appleicloud","viber","appleitunes","radius","windowsupdate","teamviewer","tuenti","lotusnotes",\
+"sap","gtp","upnp","llmnr","remotescan","spotify","h323","openvpn","noe","ciscovpn","teamspeak","tor","tor","rtcp","rsync","oracle","corba","ubuntuone"
+
+
 #define NDPI_DETECTION_SUPPORT_IPV6
 #define NDPI_PROTOCOL_HISTORY_SIZE				3
 
