Index: iptables-mod-ndpi-1.2/src/lib/ndpi_main.c
===================================================================
--- iptables-mod-ndpi-1.2.orig/src/lib/ndpi_main.c
+++ iptables-mod-ndpi-1.2/src/lib/ndpi_main.c
@@ -2741,32 +2741,6 @@ void ndpi_set_protocol_detection_bitmask
   }
 #endif
 
-#ifdef NDPI_PROTOCOL_LOTUS_NOTES
-  if (NDPI_COMPARE_PROTOCOL_TO_BITMASK(*detection_bitmask, NDPI_PROTOCOL_LOTUS_NOTES) != 0) {
-    ndpi_struct->callback_buffer[a].func = ndpi_search_lotus_notes;
-    ndpi_struct->callback_buffer[a].ndpi_selection_bitmask =
-      NDPI_SELECTION_BITMASK_PROTOCOL_V4_V6_TCP_WITH_PAYLOAD_WITHOUT_RETRANSMISSION;
-
-    NDPI_SAVE_AS_BITMASK(ndpi_struct->callback_buffer[a].detection_bitmask, NDPI_PROTOCOL_UNKNOWN);
-    NDPI_ADD_PROTOCOL_TO_BITMASK(ndpi_struct->callback_buffer[a].detection_bitmask, NDPI_PROTOCOL_LOTUS_NOTES);
-    NDPI_SAVE_AS_BITMASK(ndpi_struct->callback_buffer[a].excluded_protocol_bitmask, NDPI_PROTOCOL_LOTUS_NOTES);
-    a++;
-  }
-#endif
-
-#ifdef NDPI_PROTOCOL_GTP
-  if (NDPI_COMPARE_PROTOCOL_TO_BITMASK(*detection_bitmask, NDPI_PROTOCOL_GTP) != 0) {
-    ndpi_struct->callback_buffer[a].func = ndpi_search_gtp;
-    ndpi_struct->callback_buffer[a].ndpi_selection_bitmask =
-      NDPI_SELECTION_BITMASK_PROTOCOL_V4_V6_TCP_WITH_PAYLOAD_WITHOUT_RETRANSMISSION;
-
-    NDPI_SAVE_AS_BITMASK(ndpi_struct->callback_buffer[a].detection_bitmask, NDPI_PROTOCOL_UNKNOWN);
-    NDPI_ADD_PROTOCOL_TO_BITMASK(ndpi_struct->callback_buffer[a].detection_bitmask, NDPI_PROTOCOL_GTP);
-    NDPI_SAVE_AS_BITMASK(ndpi_struct->callback_buffer[a].excluded_protocol_bitmask, NDPI_PROTOCOL_GTP);
-    a++;
-  }
-#endif
-
 #ifdef NDPI_PROTOCOL_DCERPC
   if (NDPI_COMPARE_PROTOCOL_TO_BITMASK(*detection_bitmask, NDPI_PROTOCOL_DCERPC) != 0) {
     ndpi_struct->callback_buffer[a].func = ndpi_search_dcerpc;
