Index: iptables-mod-ndpi-1.2/src/lib/protocols/http.c
===================================================================
--- iptables-mod-ndpi-1.2.orig/src/lib/protocols/http.c
+++ iptables-mod-ndpi-1.2/src/lib/protocols/http.c
@@ -27,6 +27,8 @@
 
 #ifdef NDPI_PROTOCOL_HTTP
 
+static char *ndpi_strdup(const char *s);
+
 static void ndpi_int_http_add_connection(struct ndpi_detection_module_struct *ndpi_struct, 
 					 struct ndpi_flow_struct *flow,
 					 u_int32_t protocol)
@@ -552,7 +554,7 @@ void ndpi_http_subprotocol_conf(struct n
   }
   host_match = tmp_host_match; 
   
-  host_match[host_match_num_items].string_to_match = strdup(value);
+  host_match[host_match_num_items].string_to_match = ndpi_strdup(value);
   if (host_match[host_match_num_items].string_to_match == NULL) {
     return;
   }
