From 783e1cf322bce051a64c5d0ee9e19347c0625ee3 Mon Sep 17 00:00:00 2001
From: Willy Tarreau <w@1wt.eu>
Date: Tue, 17 Dec 2013 11:11:16 +0100
Subject: [PATCH 1/2] BUILD: last release inadvertently prepended a "+" in
 front of the date

No impact.
Index: haproxy-1.5-dev21/VERDATE
===================================================================
--- haproxy-1.5-dev21.orig/VERDATE
+++ haproxy-1.5-dev21/VERDATE
@@ -1,2 +1 @@
-+2013-12-17 00:45:49 +0100
 2013/12/17
