From 1432a0c5db8af09f51ce2bd75f3a8ecda266aee9 Mon Sep 17 00:00:00 2001
From: Thierry FOURNIER <tfournier@exceliance.fr>
Date: Tue, 11 Mar 2014 13:42:38 +0100
Subject: [PATCH 063/103] MINOR: dumpstats: Group map inline help

---
 src/dumpstats.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/src/dumpstats.c b/src/dumpstats.c
index 0b30585..afd333c 100644
--- a/src/dumpstats.c
+++ b/src/dumpstats.c
@@ -131,7 +131,6 @@ static const char stats_sock_usage_msg[] =
 	"Unknown command. Please enter one of the following commands only :\n"
 	"  clear counters : clear max statistics counters (add 'all' for all counters)\n"
 	"  clear table    : remove an entry from a table\n"
-	"  clear map [id] : clear the content of this map\n"
 	"  help           : this message\n"
 	"  prompt         : toggle interactive mode with prompt\n"
 	"  quit           : disconnect\n"
@@ -141,19 +140,21 @@ static const char stats_sock_usage_msg[] =
 	"  show errors    : report last request and response errors for each proxy\n"
 	"  show sess [id] : report the list of current sessions or dump this session\n"
 	"  show table [id]: report table usage stats or dump this table's contents\n"
-	"  show map [id]  : report avalaible maps or dump this map's contents\n"
 	"  get weight     : report a server's current weight\n"
 	"  set weight     : change a server's weight\n"
 	"  set table [id] : update or create a table entry's data\n"
 	"  set timeout    : change a timeout setting\n"
 	"  set maxconn    : change a maxconn setting\n"
 	"  set rate-limit : change a rate limiting value\n"
-	"  set map [id] [key] [value] : modify map entry\n"
-	"  add map [id] [key] [value] : add map entry\n"
-	"  del map [id] [key] : delete map entry\n"
 	"  disable        : put a server or frontend in maintenance mode\n"
 	"  enable         : re-enable a server or frontend which is in maintenance mode\n"
 	"  shutdown       : kill a session or a frontend (eg:to release listening ports)\n"
+	"  show map [id]  : report avalaible maps or dump a map's contents\n"
+	"  get map        : reports the keys and values matching a sample for a map\n"
+	"  set map        : modify map entry\n"
+	"  add map        : add map entry\n"
+	"  del map        : delete map entry\n"
+	"  clear map <id> : clear the content of this map\n"
 	"";
 
 static const char stats_permission_denied_msg[] =
-- 
1.8.3.2

